content/issue-2/applicative-parsing.md

```diff
@@ -0,0 +1,675 @@
+# Programming is hard, looking for better abstractions
+
+If you ever wrote a program bigger than a few lines of code you probably know that the bigger
+the program gets - the harder it's to make sure it does exactly what you meant due to how
+different parts can interact. Over time programmers and computer scientists discovered multiple
+ways to manage this complexity. Programming went from physically wiring the instructions to
+writing them as mnemonics with assembly language, to abstractions with variables, labels,
+functions, and so on.
+
+
+[Functional Programming](https://en.wikipedia.org/wiki/Functional_programming?useskin=vector)
+is one of the paradigms for managing complexity with its main idea of writing code as a set of
+functions and using those functions to pass values around instead imperative style of steps
+to perform to mutate some data.
+
+Rust supports both functional and imperative approaches well and parts of its standard library
+are implemented with functional use in mind. Following examples show the difference between two
+approaches:
+
+Imperative approach describes exact steps program needs to take, for example, to calculate a
+sum of all the vector items:
+- initialize `sum` with 0
+- start at the first element of the array
+- multiply `i`th element of the array by two and add it to `sum`
+- proceed to the next element
+- repeat until the last element of the array
+
+```rust
+let xs = vec![1,2,3,4];
+let mut sum = 0;
+let mut i = 0;
+loop {
+    sum += xs[i] * 2;
+    i += 1;
+    if i == xs.len() {
+        break;
+    }
+}
+```
+
+While functional approach describes what the result is:
+- start from a sequence containing all the items in the array
+- make a new sequence from it by doubling all the items
+- result is a sum of all the elements of this new sequence
+
+```rust
+let xs = vec![1,2,3,4];
+let sum = xs.iter().map(|x| x * 2).sum();
+```
+
+Both examples compile to the same set of instructions but it is easier to make a mistake in the
+imperative one.
+
+Not all the tasks are perfect fit for functional approach, at least not in an obvious way. This
+article is going to explore ways of using purely functional approach for a problem that seems
+to be better suited for imperative style.
```

Comment:
The way I see it functional approach means lack of mutability in external API and high level composition as one of the main tools. With parsing libraries I worked with you usually get either something declarative that can't do much outside of the explicitly supported cases (most of the cli parsers I looked at), something very imperative (some cli parsers) or something that exposes internal details too much and makes you to thread the state around and be very careful about what's validated and what isn't - mostly nom.

Reply:
^ That would be a way better introduction to motivate an generic applicative parser ;)

Reply:
You can totally use the approach described here to make applicative and monadic parsers in other areas. I have one more to parse TUI style input. Not published properly but [visible](https://github.com/pacak/omnomnomicon) on github in his nom-style state and monadic style is being worked on.

Making it fully generic is harder because HKTs are not thing in Rust :(

Reply:
Sure but that's a "What else" section then.. kind of

1. why bother writing an applicative parser for CLI
2. how to
3. look works great and can parse basically anything, so what else can we apply this concept to...?

There are not a whole much of this lets say haskell inspired functional approaches to things in rust so far. The ecosystems are related but haskell is just way more flexible with its more powerful type system.
It does come with some nice approaches to things stemming from more theoretical backgrounds and this article is great to show how to make use of those in rust.
An extended version could go into the limitations of rust HKT and such but i feel that has been talked about in length elsewhere.